designs/2021-schema-object-rules/README.md

```diff
@@ -0,0 +1,287 @@
+- Repo: eslint/eslint and @eslint/eslintrc
+- Start Date: 2021-12-12
+- RFC PR: <https://github.com/eslint/rfcs/pull/85>
+- Authors: [bmish](https://github.com/bmish)
+
+# Require schemas and object-style rules
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+
+We propose two changes to rules:
+
+- Requiring rules with options to specify schemas (removing support for rules with options that are missing schemas)
+- Requiring rule implementations to use the object-style format (removing support for the deprecated function-style format)
+
+Note that while these two proposals are included in the same RFC so that they can be considered holistically, they do not have to be tied together, and can be split into separate RFCs if desired.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+### Motivation for requiring schemas
+
+Today, rule schemas ([meta.schema](https://eslint.org/docs/developer-guide/working-with-rules#rule-basics)) are optional. This has the following effects:
+
+- If a rule does not specify a schema, nothing is enforced.
+- Rules with options may not have a schema specified, despite it being best practice to specify a schema for rule options. This increases the chances of developers passing invalid rule options and encountering unexpected behavior / bugs in rules.
+- If a rule author wants to enforce that no options are passed to their rule, they have to manually specify `schema: []`, otherwise it's possible that developers could accidentally provide useless rule options to their rule without knowing it.
```

Comment:
I believe `schema: {}` works as well?

Reply:
I just tested to confirm that `schema: {}` does not actually enforce anything. As mentioned in comments like https://github.com/eslint/rfcs/pull/85#discussion_r809932236, we should probably disallow no-op schemas like that. If someone wants to enforce that there are no options, they should omit the schema (after this RFC) or use `schema: []`. If someone wants to skip writing a schema, they will be able to use `schema: false`.